File: Sources/FocusCore/FocusScriptBuilder.swift
import Foundation

public enum FocusScriptBuilder {
    public static func makeUserScript(
        for phase: FocusPageRule.RunPhase,
        settings: FocusSettings,
        rules: [FocusPageRule] = FocusRuleCatalog.defaultRules
    ) throws -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]

        let configJSON = try encode(settings, with: encoder)
        let rulesJSON = try encode(rules, with: encoder)

        return FocusScriptSource.runtimeTemplate
            .replacingOccurrences(of: "__FOCUS_CONFIG_JSON__", with: configJSON)
            .replacingOccurrences(of: "__FOCUS_RULES_JSON__", with: rulesJSON)
            .replacingOccurrences(of: "__FOCUS_PHASE__", with: phase.rawValue)
    }

    private static func encode<T: Encodable>(_ value: T, with encoder: JSONEncoder) throws -> String {
        let data = try encoder.encode(value)
        return String(decoding: data, as: UTF8.self)
    }
}

File: focus-core/src/commonMain/kotlin/org/bilibilifocus/core/injection/FocusScriptBuilder.kt
package org.bilibilifocus.core.injection

import kotlinx.serialization.json.Json
import org.bilibilifocus.core.model.FocusPageRule
import org.bilibilifocus.core.model.FocusSettings

object FocusScriptBuilder {
    private val json = Json {
        prettyPrint = false
        encodeDefaults = true
    }

    private val runtimeTemplate: String by lazy {
        readBundleResource("focus-runtime.js")
    }

    fun makeUserScript(
        phase: FocusPageRule.RunPhase,
        settings: FocusSettings,
        rules: List<FocusPageRule> = FocusRuleCatalog.defaultRules,
    ): String {
        val configJSON = json.encodeToString(FocusSettings.serializer(), settings)
        val rulesJSON = json.encodeToString(
            kotlinx.serialization.builtins.ListSerializer(FocusPageRule.serializer()),
            rules,
        )

        return runtimeTemplate
            .replace("__FOCUS_CONFIG_JSON__", configJSON)
            .replace("__FOCUS_RULES_JSON__", rulesJSON)
            .replace(
                "__FOCUS_PHASE__",
                when (phase) {
                    FocusPageRule.RunPhase.DOCUMENT_START -> "documentStart"
                    FocusPageRule.RunPhase.DOCUMENT_END -> "documentEnd"
                },
            )
    }
}
